pkg/sos/objectstorage.go

```diff
@@ -0,0 +1,209 @@
+	exoscaleClient    *egoscale.Client
+	database          *db.Database
+	bucketDetails     []BucketDetail
+	aggregatedBuckets map[string]db.AggregatedBucket
+	billingDate       time.Time
+}
+
+type BucketDetail struct {
+	Organization, BucketName, Namespace string
+}
+
+func NewObjectStorage(exoscaleClient *egoscale.Client, k8sClient *k8s.Client, databaseURL string) ObjectStorage {
+	return ObjectStorage{
+		exoscaleClient: exoscaleClient,
+		k8sClient:      k8sClient,
+		database:       &db.Database{Url: databaseURL},
+	}
+}
+
+func (o *ObjectStorage) Execute(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Running metrics collector by step")
+
+	p := pipeline.NewPipeline[context.Context]()
+	p.WithSteps(
+		p.NewStep("Fetch managed buckets", o.fetchManagedBuckets),
+		p.NewStep("Get bucket usage", o.getBucketUsage),
+		p.NewStep("Get billing date", o.getBillingDate),
+		p.NewStep("Save to database", o.saveToDatabase),
+	)
+	return p.RunWithContext(ctx)
+}
+
+// getBucketUsage gets bucket usage from Exoscale and matches them with the bucket from the cluster
+// If there are no buckets in Exoscale, the API will return an empty slice
+func (o *ObjectStorage) getBucketUsage(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Fetching bucket usage from Exoscale")
+	resp, err := o.exoscaleClient.ListSosBucketsUsageWithResponse(ctx)
+	if err != nil {
+		return err
+	}
+
+	aggregatedBuckets := getAggregatedBuckets(ctx, *resp.JSON200.SosBucketsUsage, o.bucketDetails)
+	if len(aggregatedBuckets) == 0 {
+		log.Info("There are no bucket usage to be saved in the database")
+		return nil
+	}
+
+	o.aggregatedBuckets = aggregatedBuckets
+	return nil
+}
+
+func (o *ObjectStorage) fetchManagedBuckets(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Fetching buckets from cluster")
+
+	buckets := exoscalev1.BucketList{}
+	client := *o.k8sClient
+	log.V(1).Info("Listing buckets from cluster")
+	err := client.List(ctx, &buckets)
+	if err != nil {
+		return fmt.Errorf("cannot list buckets: %w", err)
+	}
+	o.bucketDetails = addOrgAndNamespaceToBucket(ctx, buckets)
+	return nil
+}
+
+func (o *ObjectStorage) saveToDatabase(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Creating a database connection")
+
+	log.V(1).Info("Opening database connection")
+	err := o.database.OpenConnection()
+	if err != nil {
+		return err
+	}
+	defer o.database.CloseConnection()
+
+	log.V(1).Info("Ensuring initial database configuration")
+	err = o.database.EnsureInitConfiguration(ctx)
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("Saving buckets information usage to database")
+	for namespace, aggregatedBucket := range o.aggregatedBuckets {
+		err = o.database.EnsureBucketUsage(ctx, namespace, aggregatedBucket, o.billingDate)
+		if err != nil {
+			log.Info("WARNING: Buckets usage have not been saved to the database", "namespace", namespace, "error", err.Error())
+		}
```

Comment:
> Wouldn't it be better if EnsureBucketUsage handles the error and logs a warning if there's nothing to save?
What is the difference with how it is implemented right now? Either log is added inside `EnsureBucketUsage` or outside of `EnsureBucketUsage` the output is the same. My idea behind this error handling is to not stop the application in case one record was no saved.  If this is not the optimal solution how would you do it?

Reply:
@Kidswiss I think we discussed this particular issue. We decided to leave the application running in case a bucket is not saved to the DB. Now what if all the buckets are not persisted, @ccremer is right to say that this will most probably go unchecked. How do we want to proceed here? throw an error right away?

Reply:
@ZuGao wasn't our discussion about cases, when the value for the buckets is 0? I'm not sure anymore...

But the architecture has changed quite a bit since our last discussion. Instead of handling multiple clusters per run, it now only handles one. So aborting the whole job seems more viable now, if values can't be saved.

Reply:
I propose to figure out the exact condition when it's ok that data wasn't saved. Maybe it's already there, maybe no bucket matched etc. Only then ignore an error, but otherwise everything else should fail the job IMO.

Reply:
The only time that it is not saving at that point is when the data already is there (I suppose it's overrides the existing row in the table). I have to check it out if I can figure out if the data is already there in the database without changing the library I'm using from cloudscale repository. Also this case should not be considered as an error but more of a warning. The rest of the cases it will return an error if it wasn't saved therefore the application should stop.

Reply:
As I expected the library does first a get to check if the record exists then it will insert if it cannot find anything. Unfortunately it does not log anything and will have to change that in the source code there. I will leave it as is here instead of doing another get before that.